File: plugins/plugin-code/typescript/actions/editFile.ts
import type {
  Action,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import type { CoderService } from "../services/coderService";

function getInputs(
  options: HandlerOptions | undefined,
  message: Memory,
): { filepath: string; oldStr: string; newStr: string } {
  const opt = options as
    | { filepath?: string; old_str?: string; new_str?: string }
    | undefined;
  const filepath = opt?.filepath?.trim() ?? "";
  const oldStr = opt?.old_str ?? "";
  const newStr = opt?.new_str ?? "";
  if (filepath) return { filepath, oldStr, newStr };

  const text = message.content.text ?? "";
  const m = text.match(/["'`]([^"'`]+)["'`]/);
  const inferred = m?.[1]?.trim() ?? "";
  return { filepath: inferred, oldStr, newStr };
}

const spec = requireActionSpec("EDIT_FILE");

export const editFile: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,
  validate: async (runtime: IAgentRuntime): Promise<boolean> =>
    runtime.getService<CoderService>("coder") !== null,
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | undefined,
    options: HandlerOptions | undefined,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = runtime.getService<CoderService>("coder");
    if (!svc) {
      const msg = "Coder service is not available.";
      if (callback) await callback({ content: { text: msg } });
      return { success: false, text: msg };
    }

    const { filepath, oldStr, newStr } = getInputs(options, message);
    if (!filepath || oldStr.length === 0) {
      const msg = "Missing filepath or old_str.";
      if (callback) await callback({ content: { text: msg } });
      return { success: false, text: msg };
    }

    const conv = message.roomId ?? message.agentId ?? runtime.agentId;
    const result = await svc.editFile(conv, filepath, oldStr, newStr);
    if (!result.ok) {
      if (callback) await callback({ content: { text: result.error } });
      return { success: false, text: result.error };
    }

    const msg = `Edited ${filepath}`;
    if (callback) await callback({ content: { text: msg } });
    return { success: true, text: msg };
  },
};

File: plugins/plugin-code/rust/src/actions/edit_file.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::service::CoderService;
use crate::{Action, ActionExample, ActionResult};

pub struct EditFileAction;

#[async_trait]
impl Action for EditFileAction {
    fn name(&self) -> &str {
        "EDIT_FILE"
    }

    fn similes(&self) -> Vec<&str> {
        vec!["REPLACE_IN_FILE", "PATCH_FILE", "MODIFY_FILE"]
    }

    fn description(&self) -> &str {
        "Replace a substring in a file (single replacement)."
    }

    async fn validate(&self, _message: &Value, _state: &Value) -> bool {
        true
    }

    async fn handler(
        &self,
        message: &Value,
        state: &Value,
        service: Option<&mut CoderService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "Coder service is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        let filepath = state.get("filepath").and_then(|v| v.as_str()).unwrap_or("");
        let old_str = state.get("old_str").and_then(|v| v.as_str()).unwrap_or("");
        let new_str = state.get("new_str").and_then(|v| v.as_str()).unwrap_or("");

        if filepath.trim().is_empty() || old_str.is_empty() {
            return ActionResult {
                success: false,
                text: "Missing filepath or old_str.".to_string(),
                data: None,
                error: Some("missing_args".to_string()),
            };
        }

        let conv = message
            .get("room_id")
            .and_then(|v| v.as_str())
            .or_else(|| message.get("agent_id").and_then(|v| v.as_str()))
            .unwrap_or("default");

        match svc.edit_file(conv, filepath, old_str, new_str).await {
            Ok(()) => ActionResult {
                success: true,
                text: format!("Edited {}", filepath),
                data: None,
                error: None,
            },
            Err(err) => ActionResult {
                success: false,
                text: err,
                data: None,
                error: Some("edit_failed".to_string()),
            },
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "edit file".to_string(),
            agent_response: "Editing…".to_string(),
        }]
    }
}
